### PR TITLE
Remove explicit comparisons to `nullptr` and most usage of `has_value()`

### PR DESCRIPTION
**Pull request by klausler**
There seems to be a consensus that `if (p != nullptr)` and `if (p.has_value())` would be more readable if they were more simply written `if (p)`.

### CODE REVIEW

File: lib/parser/token-parsers.h [hunk 1]
```diff
@@ -289,7 +289,7 @@ struct BOZLiteral {
     space.Parse(state);
     const char *start{state.GetLocation()};
     std::optional<const char *> at{nextCh.Parse(state)};
-    if (!at.has_value()) {
+    if (!at) {
```
Comment (tskeith) [inline lib/parser/token-parsers.h]: I'm seeing a bunch of compilation errors with GCC 8.3.0. For example, on this line:
```
../../lib/parser/token-parsers.h:292:9: error: could not convert 'Fortran::parser::operator!<std::optional<const char*> >(at)' from 'Fortran::parser::NegatedParser<std::optional<const char*> >' to 'bool'
```
It doesn't complain about line 143, which looks the same.

Comment (klausler) [reply lib/parser/token-parsers.h]: Already fixed; will not merge until all builds and tests are clean.

Comment (tskeith) [reply lib/parser/token-parsers.h]: What was wrong? It looks okay to me.

Comment (klausler) [reply lib/parser/token-parsers.h]: `operator!()` was already overloaded in the context of parser combinators.

**Pull request merged**